### PR TITLE
Add WitnessCS, SizedWitness, and related ConstraintSystem methods.

### PR DESCRIPTION
**Pull request by porcuquine**
This PR adds some new ConstraintSystem methods. These methods are intended to only be implemented by or called on ConstraintSystems that are also designated as witness generators. Such ConstraintSystems will signal this by implementing and returning true from `is_witness_generator()`. 

A minimal `WitnessCS` suitable for this purpose is also provided, as is a `SizedWitness` trait which should be implemented by objects (for example, implementers of `Circuit`) that choose to offer efficient witness-generation.

In this context, 'efficient witness generation' likely bypasses the usual, flexible mechanism used to specify `bellperson` circuits. Although convenient, this mechanism conflates synthesis and witness generation in ways that can add significant overhead.

Initial synthesis benchmarks in Neptune, although imperfect, showed 15x speedup when using this method, and this PR's immediate purpose is to support that use case.

### CODE REVIEW

File: src/constraint_system.rs [hunk 1]
```diff
@@ -156,6 +156,44 @@ pub trait ConstraintSystem<Scalar: PrimeField>: Sized + Send {
             "ConstraintSystem::extend must be implemented for types implementing ConstraintSystem"
         );
     }
+
+    /// ConstraintSystems that are witness generators need not assemble the actual constraints. Rather, they exist only
+    /// to efficiently create a witness.
+    fn is_witness_generator(&mut self) -> bool {
+        false
+    }
```
Comment (huitseeker) [inline src/constraint_system.rs:164]: Why does this take `&mut self` if this is a flag?

File: src/util_cs/test_cs.rs [hunk 2]
```diff
@@ -176,6 +176,17 @@ impl<Scalar: PrimeField> TestConstraintSystem<Scalar> {
...
+        self.inputs
+            .iter()
+            .map(|(scalar, _string)| *scalar)
+            .collect()
+    }
+
+    pub fn scalar_aux(&self) -> Vec<Scalar> {
+        self.aux.iter().map(|(scalar, _string)| *scalar).collect()
+    }
+
```
Comment (huitseeker) [inline src/util_cs/test_cs.rs:189]: This (and their alter ego in `WitnessCS`) seems orthogonal from the thrust of the PR, am I missing something?

File: src/util_cs/witness_cs.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+impl<Scalar> PartialEq for WitnessCS<Scalar>
+where
+    Scalar: PrimeField,
+{
+    fn eq(&self, other: &WitnessCS<Scalar>) -> bool {
+        self.input_assignment == other.input_assignment
+            && self.aux_assignment == other.aux_assignment
+    }
+}
```
Comment (huitseeker) [inline src/util_cs/witness_cs.rs:75]: `cargo-expand` reveals this is identical to the implementations generated by derive, is there a particular reason to not use them?

**Review by huitseeker — COMMENTED** [review]
Preserving the invariant that public `ConstraintSystem` methods have rustdoc would probably help convey the intent of witness constraint systems in a more persistent manner than the present PR summary.

File: src/constraint_system.rs [hunk 4]
```diff
@@ -222,6 +260,22 @@ impl<'cs, Scalar: PrimeField, CS: ConstraintSystem<Scalar>> ConstraintSystem<Sca
     fn get_root(&mut self) -> &mut Self::Root {
         self.0.get_root()
     }
+
+    fn is_witness_generator(&mut self) -> bool {
+        self.get_root().is_witness_generator()
```
Comment (huitseeker) [inline src/constraint_system.rs:265]: How about `self.0.is_witness_generator()`?

File: src/constraint_system.rs (hunk 1 above)
Comment (porcuquine) [reply src/constraint_system.rs:164]: For some reason, using `&self` everywhere leads to a compilation error. It's unclear to me what's going on.
```
error[E0596]: cannot borrow `*self` as mutable, as it is behind a `&` reference                                                                                                            |    fn is_witness_generator(&self) -> bool {                                                                                                                                                
   --> src/constraint_system.rs:265:9                                                                                                                                                      |        false                                                                                                                                                                               
    |                                                                                                                                                                                      |    }                                                                                                                                                                                       
162 |     fn is_witness_generator(&self) -> bool {                                                                                                                                         |                                                                                                                                                                                            
    |                             ----- help: consider changing that to be a mutable reference: `&mut self`                                                                                |    fn extend_inputs(&mut self, _new_inputs: &[Scalar]) {                                                                                                                                   
...                                                                                                                                                                                        |        assert!(!self.is_witness_generator());                                                                                                                                              
265 |         self.get_root().is_witness_generator()                                                                                                                                       |        unimplemented!()                                                                                                                                                                    
    |         ^^^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable                                                                            |    }
```

File: src/util_cs/test_cs.rs (hunk 2 above)
Comment (porcuquine) [reply src/util_cs/test_cs.rs:189]: They are need for testing. For example, in the downstream [neptune PR](https://github.com/lurk-lab/neptune/pull/190):

https://github.com/lurk-lab/neptune/blob/3db8506f7ad551c683d2e20784240115ca723abe/src/circuit2_witness.rs#L444-L454

File: src/util_cs/witness_cs.rs (hunk 3 above)
Comment (porcuquine) [reply src/util_cs/witness_cs.rs:75]: Nope, the fields requiring it have been gradually chipped away from the original template (`ProvingAssignment` I believe) as this implementation gets moved around and modified.

File: src/constraint_system.rs (hunk 4 above)
Comment (porcuquine) [reply src/constraint_system.rs:265]: These express the intention and are more foolproof as written, I think. But yes, I believe your rewrite is equivalent given how namespaces are implemented.

File: src/constraint_system.rs (hunk 1 above)
Comment (huitseeker) [reply src/constraint_system.rs:164]: Fix this and it will go away: https://github.com/filecoin-project/bellperson/pull/309#discussion_r1214400489

Comment (porcuquine) [reply src/constraint_system.rs:164]: Aha!

Comment (porcuquine) [reply src/constraint_system.rs:164]: However, my joy is short-lived. Making those changes does indeed fix the type problems here. But it causes downstream tests to fail.

```
--- STDERR:              neptune circuit2_witness::test::test_poseidon_hash_witness ---
thread 'circuit2_witness::test::test_poseidon_hash_witness' panicked at 'assertion failed: cs.is_witness_generator()', src/circuit2_witness.rs:32:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This makes me believe the change of implementation is *not* actually equivalent, even though it looks like it should be. This is the kind of complicated, picky, entangled … fun I've been griping about.

I'll be happy if it turns out I'm missing something obvious, though.

File: src/constraint_system.rs (hunk 4 above)
Comment (porcuquine) [reply src/constraint_system.rs:265]: Or maybe not. https://github.com/filecoin-project/bellperson/pull/309#discussion_r1215036785

File: src/constraint_system.rs (hunk 1 above)
Comment (porcuquine) [reply src/constraint_system.rs:164]: Maybe I need to properly propagate this through to neptune. That failure is worrisome, though.

Comment (porcuquine) [reply src/constraint_system.rs:164]: Alas, even after applying your suggested changes to neptune, the error persists.

Comment (huitseeker) [reply src/constraint_system.rs:164]: What is going on, specifically, in the failure of `test_poseidon_hash_witness`, is the following:
- the first assertion (`assert!(cs.is_witness_generator())`) of [`poseidon_hash_allocated_witness`](https://github.com/lurk-lab/neptune/blob/47a96a7aef7e16caddd188178eabd2d8ea65c403/src/circuit2_witness.rs#L22-L26) fails,
- this function unavoidably[^1] is called on a `&mut wcs` where `wcs` is a `WitnessConstraintSystem` [here](https://github.com/lurk-lab/neptune/blob/47a96a7aef7e16caddd188178eabd2d8ea65c403/src/circuit2_witness.rs#L439-L440). You can of course check, *before* the call to `poseidon_hash_allocated_witness` that triggers the panic, that `wcs.is_witness_generator()` returns true,
- so the instance of `ConstraintSystem::is_witness_generator` cannot be that of the `WitnessConstraintSystem` of `wcs`, but instead it is this one, labeled as a convenience:
https://github.com/filecoin-project/bellperson/blob/37ac590ea3eab8ce76699f95e191ceaece928519/src/constraint_system.rs#L233-L288
- the reason why this returns `false` on `is_witness_generator`, in the case of our test, is that this implementation omits the `is_witness_generator` field, and therefore defers to the underlying default implementation situated in the trait definition,
- note that this also omits `is_extensible`, which is liable to exhibiting the same difference in behaviour for the flag depending on whether it is called on a value / reference, or on an `&mut`,
- the reason this would pass when the flag had the signature `is_witness_generator(&mut self)` is because this signature invites the type system prioritize a search for a method part of an instance of `ConstraintSystem` on `self`, before looking up one on (&mut self),[^2]
- While backwards-compatibility requires us to keep the `ConstraintSystem` for `&mut T`, I think its behavior is a clear bug w.r.t at least flags, and possibly for any other member where it defers to the default fields. Such a convenience implementation should not change *any* aspect of the instance of `ConstraintSystem` it is a proxy for [^3]
- So I would advise fixing this issue by making sure that convenience instance implements at least all flags (`is_extensible`, `is_witness_generator`) by delegation, 
- I would also recommend implementing the other missing members (`namespace`, `extend`, `extend_inputs`, `extend_aux`, `allocate_empty`, `allocate_empty_inputs`, `allocate_empty_aux`) and bumping a major version to indicate a change in behavior that may break backwards compatibility.

[^1]: See the signature of `poseidon_hash_witness_into_cs`
[^2]: if you must, look up 'autoref' for the priorities of method lookup, or explore the [compiler](https://github.com/rust-lang/rust/blob/master/compiler/rustc_hir_typeck/src/method/probe.rs)
[^3]: Other instances of `ConstraintSystem` that are derived from an instance of a `ConstraintSystem` can and do make other choices (e.g. `NameSpace`). But I think that may be OK, as long as we're aware of their behavior, since they don't claim to be a simple convenience that's easy to inadvertently call - as this bug shows.

**Review by huitseeker — CHANGES_REQUESTED** [review]
The failure of the `test_poseidon_hash_witness` test when the `is_witness_generator` flag is made to take its argument by shared reference reveals a bug with the convenience implementation ` impl<'cs, Scalar: PrimeField, CS: ConstraintSystem<Scalar>> ConstraintSystem<Scalar> for &'cs mut CS` - which the test inadvertently obviated when it took its argument by unique reference.

Fortunately, this test has an easy fix: making sure this implementation defers to the implementation on the base type for all fields of the trait, and not the subset it happens to implement at the moment.

Unfortunately, because the present PR precisely adds members to the trait that are not properly deferred to in this convenience implementation, it "leans into" the bug, which forces me to request changes here. I've outlined an analysis of the bug and suggested a more minimal fix (~6 lines) in an inline comment.

**Comment by porcuquine** [comment]
@huitseeker Thanks for the fantastically detailed response. What you wrote makes sense, and I've added the missing methods. It's not obvious to me how `extend()` should (or maybe even *can*) work, given its signature, so I made it explicitly unimplemented to avoid perverse behavior. If you have thoughts or a better solution, please let me know.

**Comment by huitseeker** [comment]
@porcuquine I agree `extend` is impossible to proxy in this context. The problem is actually in `ec_gpu_gen`, I opened the following PR to suggest a fix:
https://github.com/filecoin-project/ec-gpu/pull/53
For future reference, the following branch:
https://github.com/huitseeker/bellperson/tree/witness-cs-with-extend-api
and in particular its latest commit https://github.com/huitseeker/bellperson/commit/174612dd417f6f8e57bb0908f328d6a7b5fbfa74 shows this is enough to make the signature of `extend` more sensible.

**Review by huitseeker — APPROVED** [review]
Some of the behavior this PR unearthed should probably lead to opening new issues, but there's nothing more to be done as part of this PR as far as I'm concerned. Nice work!

**Review by huitseeker — APPROVED** [review]

**Review by DrPeterVanNostrand — DISMISSED** [review]

**Review by DrPeterVanNostrand — APPROVED** [review]
